index.bs

```diff
@@ -124,22 +131,61 @@ A conversion report is a [=struct=] with the following items:
 
 # Algorithms # {#algorithms}
 
-<h3 algorithm id="creating-conversion">Creating a conversion</h3>
-TODO
+<h3 algorithm id="creating-impression">Activating an impression</h3>
 
-<h3 algorithm id="creating-impression">Creating an impression</h3>
-TODO
+To <dfn>activate an impression</dfn> from an <{a}> tag,
+create a new [=impression=] struct whose items are:
+
+: [=impression/impression source=]
+:: The current top frame's [=origin=].
+: [=impression/impression data=]
+:: The result of applying [=parsing conversion data=] to the <{a/impressiondata}>
+    attribute.
+: [=impression/conversion destination=]
+:: The result of parsing <{a/conversiondestination}> as a site.
+: [=impression/reporting endpoint=]
+:: The result of [=origin/obtaining a site=] from the result of parsing
+    <{a/reportingorigin}> as an origin.
+: [=impression/expiry=]
+:: Be a timestamp <{a/impressionexpiry}> milliseconds in the future.
+: [=impression/impression time=]
+:: Be the current time.
+
+TODO: How to store the impression
+
+<h3 algorithm id="creating-a-conversion">Creating a conversion</h3>
+
+To <dfn>create a conversion</dfn> from a [=url=]:
+
+: [=conversion/conversion source=]
+:: The current top-level origin.
```

Comment:
What's "current" here? You may need to pass in a settings object to look relative to.